utils/service_discovery/abstract_config_store.py

```diff
@@ -0,0 +1,173 @@
+# std
+import logging
+import simplejson as json
+from os import path
+
+# 3p
+from urllib3.exceptions import TimeoutError
+
+# project
+from utils.checkfiles import get_check_class, get_auto_conf, get_auto_conf_images
+from utils.singleton import Singleton
+
+log = logging.getLogger(__name__)
+
+CONFIG_FROM_AUTOCONF = 'auto-configuration'
+CONFIG_FROM_TEMPLATE = 'template'
+SD_TEMPLATE_DIR = '/datadog/check_configs'
+
+
+class KeyNotFound(Exception):
+    pass
+
+
+class AbstractConfigStore(object):
+    """Singleton for config stores"""
+    __metaclass__ = Singleton
+
+    previous_config_index = None
+
+    def __init__(self, agentConfig):
+        self.client = None
+        self.agentConfig = agentConfig
+        self.settings = self._extract_settings(agentConfig)
+        self.client = self.get_client()
+        self.sd_template_dir = agentConfig.get('sd_template_dir')
+        self.AUTO_CONF_IMAGES = get_auto_conf_images(agentConfig)
+
+    @classmethod
+    def _drop(cls):
+        """Drop the config store instance"""
+        del cls._instances[cls]
+
+    def _extract_settings(self, config):
+        raise NotImplementedError()
+
+    def get_client(self, reset=False):
+        raise NotImplementedError()
+
+    def client_read(self, path, **kwargs):
+        raise NotImplementedError()
+
+    def dump_directory(self, path, **kwargs):
+        raise NotImplementedError()
+
+    def _get_auto_config(self, image_name):
+        if image_name in self.AUTO_CONF_IMAGES:
+            check_name = self.AUTO_CONF_IMAGES[image_name]
+
+            # get the check class to verify it matches
+            check = get_check_class(self.agentConfig, check_name)
+            if check is None:
+                log.info("Could not find an auto configuration template for %s."
+                         " Leaving it unconfigured." % image_name)
+                return None
+
+            auto_conf = get_auto_conf(self.agentConfig, check_name)
+            init_config, instances = auto_conf.get('init_config'), auto_conf.get('instances')
+
+            # stringify the dict to be consistent with what comes from the config stores
```

Comment:
This could be improved, we shouldn't need to load yaml dump to json and reload the json